File: src/jvmTest/java/org/snakeyaml/engine/issues/issue36/EmitCommentTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.issues.issue36;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.snakeyaml.engine.v2.api.DumpSettings;
import org.snakeyaml.engine.v2.api.StreamDataWriter;
import org.snakeyaml.engine.v2.comments.CommentType;
import org.snakeyaml.engine.v2.common.ScalarStyle;
import org.snakeyaml.engine.v2.emitter.Emitter;
import org.snakeyaml.engine.v2.events.*;

import java.io.StringWriter;
import java.util.HashMap;

import static org.junit.jupiter.api.Assertions.assertEquals;

@org.junit.jupiter.api.Tag("fast")
public class EmitCommentTest {

  @Test
  @DisplayName("Issue 36: comment with scalar should not be ignored")
  void emitCommentWithEvent() {
    DumpSettings settings = DumpSettings.builder().setDumpComments(true).build();
    StreamDataWriter writer = new StreamToStringWriter();
    Emitter emitter = new Emitter(settings, writer);
    emitter.emit(new StreamStartEvent());
    emitter.emit(new DocumentStartEvent(false, null, new HashMap<>()));
    emitter.emit(
        new CommentEvent(CommentType.BLOCK, "Hello world!", null, null));
    emitter.emit(new ScalarEvent(null, null, new ImplicitTuple(true, true),
        "This is the scalar", ScalarStyle.DOUBLE_QUOTED));
    emitter.emit(new DocumentEndEvent(false));
    emitter.emit(new StreamEndEvent());

    assertEquals("#Hello world!\n" + "\"This is the scalar\"\n", writer.toString());
  }

  @Test
  @DisplayName("Issue 36: only comment should not be ignored")
  void emitComment() {
    DumpSettings settings = DumpSettings.builder().setDumpComments(true).build();
    StreamDataWriter writer = new StreamToStringWriter();
    Emitter emitter = new Emitter(settings, writer);
    emitter.emit(new StreamStartEvent());
    emitter.emit(new DocumentStartEvent(false, null, new HashMap<>()));
    emitter.emit(
        new CommentEvent(CommentType.BLOCK, "Hello world!", null, null));
    emitter.emit(new DocumentEndEvent(false));
    emitter.emit(new StreamEndEvent());

    assertEquals("#Hello world!\n", writer.toString());
  }
}


class StreamToStringWriter extends StringWriter implements StreamDataWriter {

}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/issues/issue36/EmitCommentTest.kt
package it.krzeminski.snakeyaml.engine.kmp.issues.issue36

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import it.krzeminski.snakeyaml.engine.kmp.api.DumpSettings
import it.krzeminski.snakeyaml.engine.kmp.api.StreamDataWriter
import it.krzeminski.snakeyaml.engine.kmp.comments.CommentType
import it.krzeminski.snakeyaml.engine.kmp.common.ScalarStyle
import it.krzeminski.snakeyaml.engine.kmp.emitter.Emitter
import it.krzeminski.snakeyaml.engine.kmp.events.*

class EmitCommentTest : FunSpec({
    test("comment with scalar should not be ignored") {
        val settings = DumpSettings(dumpComments = true)
        val writer = StreamToStringWriter()
        val emitter = Emitter(settings, writer)
        emitter.emit(StreamStartEvent())
        emitter.emit(DocumentStartEvent(explicit = false, specVersion = null, tags = emptyMap()))
        emitter.emit(CommentEvent(
            commentType = CommentType.BLOCK,
            value = "Hello world!",
            startMark = null,
            endMark = null
        ))
        emitter.emit(ScalarEvent(
            anchor = null,
            tag = null,
            implicit = ImplicitTuple(plain = true, nonPlain = true),
            value = "This is the scalar",
            scalarStyle = ScalarStyle.DOUBLE_QUOTED,
        ))
        emitter.emit(DocumentEndEvent(false))
        emitter.emit(StreamEndEvent())

        writer.toString() shouldBe "#Hello world!\n\"This is the scalar\"\n"
    }

    test("only comment should not be ignored") {
        val settings = DumpSettings(dumpComments = true)
        val writer = StreamToStringWriter()
        val emitter = Emitter(settings, writer)
        emitter.emit(StreamStartEvent())
        emitter.emit(DocumentStartEvent(explicit = false, specVersion = null, tags = emptyMap()))
        emitter.emit(CommentEvent(
            commentType = CommentType.BLOCK,
            value = "Hello world!",
            startMark = null,
            endMark = null
        ))
        emitter.emit(DocumentEndEvent(isExplicit = false))
        emitter.emit(StreamEndEvent())

        writer.toString() shouldBe "#Hello world!\n"
    }
})

private class StreamToStringWriter : StreamDataWriter {
    private val content = StringBuilder()

    override fun write(str: String) {
        content.append(str)
    }

    override fun write(str: String, off: Int, len: Int) {
        content.append(str.substring(off, off + len))
    }

    override fun flush() {
        // No-op
    }

    override fun toString(): String = content.toString()
}
